docs(test-runner): document configuration options (#7572)
microsoft/playwright: docs/src/test-configuration.md

title: "Configuration"
---

> Playwright Test provides options to configure default `browser`, `context` and `page` fixtures, for example `headless`, `viewport` or `ignoreHTTPSErrors`. You can also record video or trace for the test or capture screenshot at the end.

Reviewer: ```suggestion
Playwright Test provides options to configure the default `browser`, `context` and `page` fixtures. For example there are options for `headless`, `viewport` and `ignoreHTTPSErrors`. You can also record a video or a trace for the test or capture a screenshot at the end.
```

});
```

## Basic options

These are commonly used options for various scenarios. You usually set them globally in [configuration file](#global-configuration).

- `baseURL` - Base URL used for all pages in the context. Allows navigating by using just the path, for example `page.goto('/settings')`.

> - `browserName` - Name of the browser that will run the tests, one of `chromium`, `firefox`, `webkit`.

Reviewer: ```suggestion
- `browserName` - Name of the browser that will run the tests, one of `chromium`, `firefox`, or `webkit`.
```

> - `bypassCSP` - Toggles bypassing page's Content-Security-Policy. Useful when CSP includes production origin.

Reviewer: ```suggestion
- `bypassCSP` - Toggles bypassing Content-Security-Policy. Useful when CSP includes the production origin.
```

- `channel` - Browser channel to use. [Learn more](./browsers.md) about different browsers and channels.

> - `headless` - Whether to run browser in headless mode.

Reviewer: ```suggestion
- `headless` - Whether to run the browser in headless mode.
```

```js js-flavor=js
// playwright.config.js
// @ts-check
const { devices } = require('playwright');

/** @type {import('@playwright/test').PlaywrightTestConfig} */
const config = {
  projects: [
    // "Pixel 4" tests use Chromium browser.
    {
      name: 'Pixel 4',
      use: {
        browserName: 'chromium',
        ...devices['Pixel 4'],
      },
    },

    // "iPhone 11" tests use WebKit browser.
    {
      name: 'iPhone 11',
      use: {
        browserName: 'webkit',
        ...devices['iPhone 11'],
      },
    },
  ],
};

module.exports = config;
```

```js js-flavor=ts
// playwright.config.ts
import { PlaywrightTestConfig } from '@playwright/test';
import { devices } from 'playwright';

const config: PlaywrightTestConfig = {
  projects: [
    // "Pixel 4" tests use Chromium browser.
    {
      name: 'Pixel 4',
      use: {
        browserName: 'chromium',
        ...devices['Pixel 4'],
      },
    },

    // "iPhone 11" tests use WebKit browser.
    {
      name: 'iPhone 11',
      use: {
        browserName: 'webkit',
        ...devices['iPhone 11'],
      },
    },
  ],
};
export default config;
```

> You can also specify options separately, instead of using predefined `devices`, or configure locale, geolocation, timezone and more:

Reviewer: ```suggestion
You specify options separately instead of using predefined devices. There are also more options such as locale, geolocation, and timezone which can be configured.
```

- `testIgnore`: Glob patterns or regular expressions that should be ignored when looking for the test files. For example, `'**/test-assets'`.
- `testMatch`: Glob patterns or regular expressions that match test files. For example, `'**/todo-tests/*.spec.ts'`. By default, Playwright Test runs `.*(test|spec)\.(js|ts|mjs)` files.
- `timeout`: Time in milliseconds given to each test.
- `webServer: { command: string, port?: number, cwd?: string, timeout?: number, env?: object }` - Launch a web server before the tests will start. It will automatically detect the port when it got printed to the stdout.
- `workers`: The maximum number of concurrent worker processes to use for parallelizing tests.

> You can specify these options in the configuration file. Note that testing options are **top-level**, do not put them into `use` section.

Reviewer: ```suggestion
You can specify these options in the configuration file. Note that testing options are **top-level**, do not put them into the `use` section.
```
